text/3389-manifest-lint.md

```diff
@@ -0,0 +1,253 @@
+inherit with `workspace = true`, we could have `[workspace.lints.<preset>]`
+which defines presets and the user could do `lints.<preset> = true`.  The user
+could then name them as they wish to avoid collision with rustc lints.
+
+Instead of the `[lints]` table being `lint = "level"`, we could organize
+it around `level = ["lint", ...]` like some other linters do (like
+[ruff](https://beta.ruff.rs/docs/configuration/)) but this works better for
+logically organizing lints, highlighting what changed in diffs, and for
+possibly adding lint-specific configuration in the future.
+
+# Prior art
+[prior-art]: #prior-art
+
+Rust
+- [cargo cranky](https://github.com/ericseppanen/cargo-cranky)
+
+Python
+- [flake8](https://flake8.pycqa.org/en/latest/user/configuration.html)
+  - Format is `level = [lint, ...]`
+- [pylint](https://github.com/PyCQA/pylint/blob/main/examples/pylintrc#L402)
+  - Format is `level = [lint, ...]` but the [config file is a reflection of the CLI](https://pylint.pycqa.org/en/latest/user_guide/configuration/index.html)
+- [ruff](https://beta.ruff.rs/docs/configuration/)
+  - Format is `level = [lint, ...]`, likely due to past precedence in ecosystem (see above)
+
+Javascript
+- [eslint](https://eslint.org/docs/latest/use/configure/rules)
+  - Format is `lint = level` / `lint = [ level, additional config ]`
+
+Go
+- [golangci-lint](https://golangci-lint.run/usage/configuration/)
+  - Format is `level = [lint, ...]`
+
+Ruby
+- [rubocop](https://docs.rubocop.org/rubocop/1.45/configuration.html)
+  - Format is `Lint: Enabled: true`
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+How should we hand rustdoc lint levels or, in the future, cargo lint levels?
+The current proposal takes all lints and passes them to rustc like `RUSTFLAGS`
+but rustdoc uses `RUSTDOCFLAGS` and cargo would use neither.  This also starts
+to get into
+[user-defined tool attributes](https://rust-lang.github.io/rfcs/2103-tool-attributes.html).
+
+Should we only apply/fingerprint lints for the appropriate tool?  For example,
+we would not include and fingerprint `clippy::` lints when running builds,
+allowing them to change without forcing a rebuild.  We likely already need to
+be tool-aware for built-in tools to handle `rustdoc::` lints (see above) so
+this isn't much more of a step.
+
+How do we allow controlling precedence between lints and lint groups?  We are
+using a TOML table with the keys as lint names which does not allow controlling
+ordering.  Even if we switched to `level = [lint, ...]`, you get a hard coded
+precedence between levels that the user can't control.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Configurable lints
+
+We can extend basic lint syntax:
+```toml
+[lints]
+cyclomatic_complexity = "allow"
+```
+to support configuration, whether for cargo or the lint tool:
+```toml
+[lints]
+cyclomatic_complexity = { level = "allow", rust-version = "1.23.0", threshold = 30 }
```

Comment:
Something else to consider would be to allow both: Defining levels with `level = ["lint", ...]` AND (`lint = "level"` OR `lint = { ... }`).

Reply:
Should we pull this into the other thread about the data model?

Reply:
Sure. I extracted it from this discussion, because I rather see this as a future addition, once lint configuration is added. So in the MVP I would go for _either_ `lint = "level"` OR `level = ["lint", ...]` and then extend this once we need it for lint configurations. (I prefer the `level = ["lint", ...]` way for the MVP).

Reply:
How does lint configuration change it?

Personally, I think we should only have one syntax so we should consider the likelihood of the future state and plan to that now.

Reply:
Yeah, thinking about this more, I think only one syntax makes more sense. 

Especially because lint configuration (from a Clippy perspective) is not about configuring a lint with a value, but configuring the tool, which will then adapt the behavior of one or many lints depending on the configuration value. So I think lint/tool configuration should be separated from the decision of the format.

The only lint-specific configuration I can see are configurations of the lint levels and how cargo should treat them, i.e. `rust-version` configuration attached to the level. IMO this is best achieved with this suggestion: https://github.com/rust-lang/rfcs/pull/3389#discussion_r1107371674

Reply:
> Especially because lint configuration (from a Clippy perspective) is not about configuring a lint with a value, but configuring the tool, which will then adapt the behavior of one or many lints depending on the configuration value

From my perspective as a user, some of the configuration is general, like `avoid-breaking-exported-api`, while others are lint specific, like `cognitive-complexity-threshold`. 

Is there a reason this model is inherently incorrect that we couldn't model it this way in the tool?  For example, past discussions of lint configuration have assumed `cyclomatic_complexity = { state = "deny", threshold = 30 }` would the way to do it, rather than setting the level for `cyclomatic_complexity` in one place and then setting the threshold in another.

Reply:
> Is there a reason this model is inherently incorrect that we couldn't model it this way in the tool?

Lint configuration in Clippy is always 1-to-many. For some/most configurations it is indeed 1-to-1. But this could just change in the future and another lint could just also pick up this configuration value and use it. 

Only being able to configure one lint at a time would make things worse IMO. E.g. for the `msrv` config you would have to add >15 lines of lints with the same `msrv` config applied.

So I would stop calling this _lint_ configuration, because _lint_ configuration doesn't really exist from a Clippy perspective. I would call it _tool_ configuration, which in the case of Clippy configures lints. And with that I wouldn't try to cover this in this RFC. Not even in a `[lints.<tool>.metadata]` section.

TL;DR: The ~~cake~~ lint configuration is a lie :cake: